ddl/generated_column.go

```diff
@@ -41,21 +41,24 @@ func verifyColumnGeneration(colName2Generation map[string]columnGenerationInDDL,
 					return errors.Trace(err)
 				}
 			} else {
-				err := errBadField.GenWithStackByArgs(depCol, "generated column function")
+				err := ErrBadField.GenWithStackByArgs(depCol, "generated column function")
 				return errors.Trace(err)
 			}
 		}
 	}
 	return nil
 }
 
-// columnNamesCover checks whether dependColNames is covered by normalColNames or not.
-// it's only for alter table add column because before alter, we can make sure that all
-// columns in table are verified already.
-func columnNamesCover(normalColNames map[string]struct{}, dependColNames map[string]struct{}) error {
-	for name := range dependColNames {
-		if _, ok := normalColNames[name]; !ok {
-			return errBadField.GenWithStackByArgs(name, "generated column function")
+// ensure all depended columns are exist.
+//
+// NOTE: this will destroy parameter `dependCols`
+func checkDependedColExist(dependCols map[string]struct{}, cols []*table.Column) error {
+	for _, col := range cols {
+		delete(dependCols, col.Name.L)
+	}
+	if len(dependCols) != 0 {
+		for arbitraryKey := range dependCols {
```

Comment:
the `for` loop is not needed, just using `dependCols[0]` is enough

Reply:
`dependCols` is a map, maybe `for` loop is the only choice

Reply:
Sure, forget that.